text/1615-cargo-platform-standards.md

```diff
@@ -0,0 +1,170 @@
+
+# Detailed design
+
+We are going to introduce new environment variables:
+```
+CARGO_BIN_DIR
+CARGO_CACHE_DIR
+CARGO_CONFIG_DIR
+```
+
+For the default values of these variables if they are set to the empty string
+or not set (which will be the common case), see below.
+
+These will be used to split the current `.cargo` (`CARGO_HOME`) directory up:
+The cached packages (`.cargo/git`, `.cargo/registry`) will go into
+`CARGO_CACHE_DIR`, binaries (`.cargo/bin`) installed by Cargo will go into
+`CARGO_BIN_DIR` and the config (`.cargo/config`) will go into
+`CARGO_CONFIG_DIR`.
+
+In order to maintain backward compatibility, the old directory locations will
+be checked if the new ones don't exist. In detail, this means:
+
+1. If any of the new variables `CARGO_BIN_DIR`, `CARGO_CACHE_DIR`,
+   `CARGO_CONFIG_DIR` are set and nonempty, use the new directory structure.
+2. Else, if there is an override for the legacy Cargo directory, using
+   `CARGO_HOME`, the directories for cache, configuration and executables are
+   placed inside this directory.
+3. Otherwise, if the Cargo-specfic platform-specific directories exist, use
+   them. What constitutes a Cargo-specific directory is laid out below, for
+   each platform.
+4. If that's not the case, check whether the legacy directory exists (`.cargo`)
+   and use it in that case.
+5. If everything else fails, create the platform-specific directories and use
+   them.
+
+This makes Cargo use platform-specific directories for new installs while
+retaining compatibility for the old directory layout. It also allows one to
+keep all Cargo related data in one place if one wishes to.
+
+## Windows
+
+We'll obtain each of the following directories using the correct API.
+
+```
+cache:    AppData\Local\Temp\Cargo
+config:   AppData\Roaming\Cargo
+binaries: AppData\Local\Programs\Cargo
+```
+
+## Unixy systems (macOS, Linux, BSDs)
+
+Here, we're following the [XDG specification](https://specifications.freedesktop.org/basedir-spec/basedir-spec-0.7.html).
+By default, if no further variables are set, this means that we'll be using the
+following subdirectories of the home directory.
+
+```
+cache:    .cache/cargo
+config    .config/cargo
+binaries: .local/bin
+```
+
+**There is currently an on-going discussion about standardizing the location of
+`.local/bin` together with a new XDG variable `XDG_BIN_HOME`. The
+implementation of this RFC should be delayed until that discussion has finished
+and use the result. This RFC will be amended with that result.**
+
+
+## Rustup
+
+Rustup will replicate Cargo's priorisation algorithm. If the results differ
+from what the executed version of Cargo will do, Rustup will add environment
+variables `CARGO_BIN_DIR`, `CARGO_CACHE_DIR`, `CARGO_CONFIG_DIR` for the new
+versions of Cargo, and add symlinks for the old versions of Cargo.
+
+
+## New subcommand
+
+Cargo (and Rustup) are going to gain a new subcommand, `cargo dirs`. It will
```

Comment:
Could this be added to the `metadata` command instead of adding a whole new command?

Reply:
Or maybe to 'rustup show' since thats usably small :).